packages/app/contexts/NetworkContext.tsx

```diff
@@ -51,56 +50,58 @@ export const NetworkContextProvider = ({
 
-  const [chainId, setChainId] = useQueryState(
-    "chainId",
-    parseAsInteger.withDefault(ChainId.GNOSIS)
-  );
+  const [searchParamsChainId, setSearchParamsChainId] =
+    useQueryState("chainId");
+  const { chains } = config.getPublicClient();
+
   const [selectedChain, setSelectedChain] = useState<WagmiChain>(gnosis);
-  const [chains, setChains] = useState<WagmiChain[] | undefined>([]);
+  const [selectedChainId, setSelectedChainId] = useState<ChainId>();
 
   const { switchNetwork } = useSwitchNetwork({
     onSuccess(data) {
-      setChainId(data.id);
+      setSearchParamsChainId(data?.id.toString());
     },
   });
 
+  // sets chainId from search params
   useEffect(() => {
-    const { chains: wagmiChains } = config.getPublicClient();
-
-    if (chains) setChains(wagmiChains);
-    if (!chainId) {
-      setChainId(ChainId.GNOSIS);
-      setSelectedChain(gnosis);
+    if (isConnected && chains) {
+      if (searchParamsChainId && switchNetwork) {
+        const isValidSearchParamsChainId = checkIsValidChainId(
+          parseInt(searchParamsChainId)
+        );
+
+        if (
+          isValidSearchParamsChainId &&
+          parseInt(searchParamsChainId) !== chain?.id
+        )
+          switchNetwork(parseInt(searchParamsChainId));
+      }
+    } else {
+      if (searchParamsChainId) {
+        const chainFromSearchParams = chains?.find(
+          (chain: any) => chain.id === parseInt(searchParamsChainId)
+        );
+
+        if (chainFromSearchParams) {
+          setSelectedChain(chainFromSearchParams);
+          setSelectedChainId(chainFromSearchParams.id);
+        }
+      } else {
+        // set default chain
+        setSelectedChain(gnosis);
+        setSelectedChainId(gnosis.id);
+      }
     }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
-
-  useEffect(() => {
-    const { chains: wagmiChains } = config.getPublicClient();
-    const chainFromParams = wagmiChains?.find((chain) => chain.id === chainId);
-
-    if (isConnected && chain?.id !== chainFromParams?.id && switchNetwork)
-      switchNetwork?.(chainId);
 
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [switchNetwork]);
+  }, [isConnected, chains, searchParamsChainId, switchNetwork]);
 
   useEffect(() => {
-    const { chains } = config.getPublicClient();
-    const isValidChainId = getIsValidChainId(chainId);
-
-    if (!isValidChainId) setChainId(ChainId.GNOSIS);
-    if (chains) {
-      const validChainId = isValidChainId ? chainId : ChainId.GNOSIS;
-      const validChain = chains.find((chain) => chain.id === validChainId);
-
-      setSelectedChain(validChain as WagmiChain);
+    if (isConnected && chain) {
+      setSelectedChain(chain);
+      setSelectedChainId(chain?.id);
     }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [chainId, selectedChain]);
-
-  useEffect(() => {
-    if (isConnected && chain) setSelectedChain(chain);
   }, [chain, isConnected]);
```

Comment:
Just thinking out loud:
This hook will trigger with similar conditions to the one above it, and they both kinda set the same values (except for the first condition that switch network)

I'm not sure if this won't bring trouble in the future

Reply:
yeah, also not so confident we need this hook, will take a second look.

Reply:
Ok, got it, we need this so anytime the chain changes it updates local provided state variables. As in the app we use NetworkProvider instead of wagmi `useNetwork` hook, so we need to update those whenever, chain changes. This is only needed when wallet is connected.